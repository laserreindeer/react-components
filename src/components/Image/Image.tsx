import { useMemo } from "react";
import { wrap } from "../../utils/wrap/wrap";
import type { ImageProperties } from "./ImageProperties";

/**
 * Image component using the wrap util.
 */
export const Image = wrap("img")<ImageProperties>(
	({ srcMap, ...properties }) => {
		const srcMapEntries = useMemo(
			() => (srcMap ? Object.entries(srcMap) : undefined),
			[srcMap]
		);

		return (
			<img
				src={
					srcMapEntries?.sort(
						// eslint-disable-next-line max-params
						([previous], [current]) =>
							parseFloat(current) - parseFloat(previous)
					)[0]?.[1]
				}
				srcSet={srcMapEntries
					?.map(([size, url]) => `${url} ${size}w`)
					.join(",")}
				{...properties}
			/>
		);
	}
);
