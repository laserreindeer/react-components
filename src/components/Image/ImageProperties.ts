/**
 * Image component properties.
 */
export type ImageProperties = {
	/** Image sizes and urls */
	srcMap: Readonly<Record<number, string>>;
};
