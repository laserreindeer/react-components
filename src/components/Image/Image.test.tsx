import { render } from "@testing-library/react";
import { Image } from "./Image";

it("matches snapshot", () => {
	const { asFragment } = render(<Image />);
	expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot srcMap", () => {
	const { asFragment } = render(
		<Image
			srcMap={{
				200: "https://placekitten.com/200/200",
				300: "https://placekitten.com/300/300",
				500: "https://placekitten.com/500/500"
			}}
		/>
	);
	expect(asFragment()).toMatchSnapshot();
});
