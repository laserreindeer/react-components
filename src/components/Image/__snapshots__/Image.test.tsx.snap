// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <img />
</DocumentFragment>
`;

exports[`matches snapshot srcMap 1`] = `
<DocumentFragment>
  <img
    src="https://placekitten.com/500/500"
    srcset="https://placekitten.com/500/500 500w,https://placekitten.com/300/300 300w,https://placekitten.com/200/200 200w"
  />
</DocumentFragment>
`;
